extensions/derpygamer2142/gpusb3.js

```diff
@@ -0,0 +1,4794 @@
+      }
+
+      // @ts-ignore
+      this.adapter = await navigator.gpu.requestAdapter();
+      if (!this.adapter) {
+        if (typeof scaffolding === "undefined")
+          alert("Failed to get WebGPU adapter!");
+        return this.throwError(
+          "AdapterGetFail",
+          "Failed to get adapter",
+          "Init",
+          "Failed to get a WebGPU adapter"
+        );
+        // throw Error("Failed to get WebGPU adapter.");
+      }
+
+      this.device = await this.adapter.requestDevice();
+
+      this.device.lost.then((info) => {
+        if (typeof scaffolding === "undefined") alert("WebGPU device lost!");
+        this.throwError("DeviceLost", info.message, "WebGPU", info);
+        this.device = null; // requestDevice will never return null so we need to account for that ourselves
+      });
+
+      // note to self: uncomment this on release
+      this.device.addEventListener("uncapturederror", (event) => {
+        this.throwError(
+          "UnclassifiedError",
+          // @ts-ignore
+          event.error.message,
+          "Unknown",
+          // @ts-ignore
+          event.error
+        ); // this is literally in the spec and the mdn docs, idk why it's complaining about event.error being undefined https://www.w3.org/TR/webgpu/#eventdef-gpudevice-uncapturederror
+      });
+    }
+
+    /**
+     * Throw an error in such a way that it can be read from the project and not break stuff
+     * @param {String} errorname The name of the error, in PascalCase
+     * @param {String} errorbody A short version of the error, punctuated
+     * @param {String} errorsource The source of the error(usually a block), in PascalCase
+     * @param {String | Object} full The errorbody, but with more detail
+     */
+    throwError(errorname, errorbody, errorsource, full) {
+      error = {
+        name: errorname ?? "Undefined. This is an error, please report it!",
+        body: errorbody ?? "Undefined. This is an error, please report it!",
+        source: errorsource ?? "Undefined. This is an error, please report it!",
+        full: full ?? "Undefined. This is an error, please report it!",
+      };
+
+      console.error(error);
+
+      Scratch.vm.runtime.startHats("gpusb3_onError");
+    }
+
+    /**
+     *
+     * @param {"buffers" | "bindGroups" | "bindGroupLayouts" | "arrayBuffers" | "views" | "textures" | "samplers"} type
+     */
+    resourceFromType(type) {
+      switch (type) {
+        case "buffers":
+          return buffers;
+        case "bindGroups":
+          return bindGroups;
+        case "bindGroupLayouts":
+          return bindGroupLayouts;
+        case "arrayBuffers":
+          return arrayBuffers;
+        case "views":
+          return views;
+        case "textures":
+          return textures;
+        case "samplers":
+          return samplers; // unused, returned anyways
+        default:
+          return null;
+      }
+    }
```

Comment:
Unless I'm wrong can you make them part of an object and do something like this

```suggestion
    resourceFromType(type) {
        return resources[type];
    }
```
If it's not possible that's okay but it might help clean it up a lot since all of the strings are exactly the same as the name of the returned values.

Reply:
that was how it was originally but a reviewer asked for it to be changed